crates/starknet-types-core/src/felt.rs

```diff
@@ -170,14 +173,47 @@ impl Felt {
         Self(self.0.pow(exponent.0.representative()))
     }
 
-    /// Modular multiplication between `self` and `rhs` modulo `p`.
+    // Implemention taken from Jonathan Lei's starknet-rs
+    // https://github.com/xJonathanLEI/starknet-rs/blob/a3a0050f80e90bd40303256a85783f4b5b18258c/starknet-crypto/src/fe_utils.rs#L20
+    /// Modular multiplication between `self` and `rhs` in modulo `p`.
     pub fn mul_mod(&self, rhs: &Self, p: &NonZeroFelt) -> Self {
-        (self * rhs).div_rem(p).1
+        let multiplicand = BigInt::from_bytes_be(num_bigint::Sign::Plus, &self.to_bytes_be());
+        let multiplier = BigInt::from_bytes_be(num_bigint::Sign::Plus, &rhs.to_bytes_be());
+        let modulus = BigInt::from_bytes_be(num_bigint::Sign::Plus, &p.0.to_bytes_be());
+
+        let result = multiplicand.mul(multiplier).mod_floor(&modulus);
+
+        let (_, buffer) = result.to_bytes_be();
+        let mut result = [0u8; 32];
+
+        result[(32 - buffer.len())..].copy_from_slice(&buffer[..]);
+
+        // safe .unwrap()
+        Felt::from_bytes_be(&result).unwrap()
```

Comment:
Why not use felts and use `div_rem` to avoid (expensive?) conversions with bytes

Reply:
Because the idea of the methods is to operate with Field Elements of mod `p` instead of the Felt that operate in the p for the Starknet PRIME.
Right now we cant make this operations with lambdaworks, so we need to use BigInts

Reply:
I agree that the performance could be better but for a first implementation I think is good, also these two methods are not the most used.
In the future I will talk with the lambdaworks team to see how we can improve the performance

Reply:
Relevant info: the cited implementation is already relying on `num_bigint`, so at the very least this should not regress performance (we'll have to test eventually of course).